wick/expression.py

```diff
@@ -495,19 +495,43 @@ def _print_str(self, with_scalar=True):
             out += tt._print_str(imap)
         return out
 
-    def _einsum_str(self):
+    def _einsum_str(self, exprs_with_space = None, names_with_space = None, space_idx_formatter = None, einsum_str_formatter = None):
         imap = self._idx_map()
         sstr = str(float(self.scalar))
         fstr = str()
         istr = str()
         tstr = str()
+
+        names_with_space = [] if names_with_space is None else names_with_space
+
+        if exprs_with_space:
+            assert space_idx_formatter is not None
+            for expr in exprs_with_space:
+                for term in expr.terms:
+                    for tt in term.tensors:
+                        names_with_space.append(tt.name)
+
         for tt in self.tensors:
             if not tt.name:
                 fstr += tt._istr(imap)
             else:
-                tstr += ", " + tt.name
+                if tt.name in names_with_space:
+                    space_idx_list = []
+
+                    for ii in tt.indices:
+                        space_idx_list.append(ii.space)
+
+                    tstr += ", " + space_idx_formatter(tt.name, space_idx_list)
+
+                else:
+                    tstr += ", " + tt.name
+
                 istr += tt._istr(imap) + ","
-        return sstr + "*einsum('" + istr[:-1] + "->" + fstr + "'" + tstr + ")"
+
+        if einsum_str_formatter is not None:
+            return einsum_str_formatter(sstr, istr, fstr, tstr)
+        else:
```

Comment:
This `else` is unnecessary, you can just have an unintended return after the the `if` statement